src/security/collection.cpp

```diff
@@ -0,0 +1,396 @@
+  static constexpr std::string_view kMethod{"server.request.method"};
+  static constexpr std::string_view kHeadersNoCookies{
+      "server.request.headers.no_cookies"};
+  static constexpr std::string_view kCookies{"server.request.cookies"};
+  static constexpr std::string_view kStatus{"server.response.status"};
+  static constexpr std::string_view kClientIp{"http.client_ip"};
+  static constexpr std::string_view kRespHeadersNoCookies{
+      "server.response.headers.no_cookies"};
+
+ public:
+  explicit ReqSerializer(dnsec::DdwafMemres &memres) : memres_{memres} {}
+
+  ddwaf_object *serialize(const ngx_http_request_t &request) {
+    dnsec::ddwaf_obj *root = memres_.allocate_objects<dnsec::ddwaf_obj>(1);
+    dnsec::ddwaf_map_obj &root_map = root->make_map(6, memres_);
+
+    set_request_query(request, root_map.at_unchecked(0));
+    set_request_uri_raw(request, root_map.at_unchecked(1));
+    set_request_method(request, root_map.at_unchecked(2));
+    set_request_headers_nocookies(request, root_map.at_unchecked(3));
+    set_request_cookie(request, root_map.at_unchecked(4));
+    set_client_ip(request, root_map.at_unchecked(5));
+
+    return root;
+  }
+
+  ddwaf_object *serialize_end(const ngx_http_request_t &request) {
+    dnsec::ddwaf_obj *root = memres_.allocate_objects<dnsec::ddwaf_obj>(1);
+    dnsec::ddwaf_map_obj &root_map = root->make_map(2, memres_);
+
+    set_response_status(request, root_map.at_unchecked(0));
+    set_response_headers_no_cookies(request, root_map.at_unchecked(1));
+
+    return root;
+  }
+
+ private:
+  static void set_map_entry_str(dnsec::ddwaf_obj &slot, std::string_view key,
+                                const ngx_str_t &value) {
+    slot.set_key(key);
+    slot.make_string(to_string_view(value));
+  }
+
+  void set_request_query(const ngx_http_request_t &request,
+                         dnsec::ddwaf_obj &slot) {
+    slot.set_key(kQuery);
+    const ngx_str_t &query = request.args;
+    if (query.len == 0) {
+      slot.make_array(nullptr, 0);
+      return;
+    }
+
+    dnsec::QueryStringIter it{query, memres_, '&',
+                              dnsec::QueryStringIter::trim_mode::no_trim};
+    set_value_from_iter(it, slot);
+  }
+
+  template <typename Iter>
+  void set_value_from_iter(Iter &it, dnsec::ddwaf_obj &slot) {
+    // first, count the number of occurrences for each key
+    std::unordered_map<std::string_view, std::size_t> keys_bag;
+    for (; !it.ended(); ++it) {
+      std::string_view key = it.cur_key();
+      // if (it.is_delete()) { }
+      // don't reduce the count, we need to allocate space for the provisional
+      // writes (before the deletion)
+      // we could improve this by saving the pointer of the first non-deleted
+      // header per key
+      keys_bag[key]++;
+    }
+
+    // we now know the number of keys; allocate map entries
+    dnsec::ddwaf_obj *entries =
+        memres_.allocate_objects<dnsec::ddwaf_obj>(keys_bag.size());
```

Comment:
I thought duplicates were tolerated, not fully supported. I mean, how do you even indicate paths of offending data? a > b` is now ambiguous if `a` contains several entries keyed `b`.

But anyway, a first pass is always necessary to avoid reallocations further down the road.